src/Prototype/Exe/Server.hs

```diff
@@ -232,8 +266,106 @@ handleLogin User.Credentials {..} =
+-- | Serve the static files for the documentation. This also provides a custom
+-- 404 fallback.
+serveDocumentation = serveDirectoryWith settings
+ where
+  settings = (defaultWebAppSettings root)
+    { ss404Handler = Just custom404
+    , ssLookupFile = webAppLookup hashFileIfExists root
+    }
+  root = "./_site/"
+
 custom404 :: Application
 custom404 _request sendResponse = sendResponse $ Wai.responseLBS
   HTTP.status404
   [("Content-Type", "text/html; charset=UTF-8")]
   (renderMarkup $ H.toMarkup Pages.NotFoundPage)
+
+
+--------------------------------------------------------------------------------
+-- Most of this is taken from WaiAppStatic.Storage.Filesystem.
+-- The goal is to allow to re-define the above `ssLookupFile` in
+-- `defaultWebAppSettings`: instead of looking up files as-is, we make it so
+-- that given a URL such as /some/path, we act as if it was /some/path.html,
+-- thus making nicer URLs (where the .html is not necessary).
+-- TODO Move this to hypered/commence.
+-- TODO Don't rewrite paths when there is an extension, otherwise it won't work
+-- for e.g. images.
+-- Note: we copy more code than necessary because it is not in the export list
+-- of the original module.
+
+-- | Construct a new path from a root and some @Pieces@.
+pathFromPieces :: FilePath -> Pieces -> FilePath
+pathFromPieces = foldl' (\fp p -> fp </> T.unpack (fromPiece p))
+
+-- | Construct a @Piece@ without input validation.
+-- The original can use a non-exported `Piece` constructor. Here we use the one
+-- that does some validation, but in an unsafe way.
+unsafeToPiece :: Text -> Piece
+unsafeToPiece t = let Just p = toPiece t in p
+
+-- | More efficient than @fileSystemLookup@ as it only concerns itself with
+-- finding files, not folders.
+-- This is the main change compared to WaiAppStatic.Storage.Filesystem:
+-- If the last piece is empty, this lookup index.html.
+-- Otherwise, this appends ".html" before trying to lookup that.
+-- This means that if you want `documentation/` to be a directory, you can
+-- decide if you want to support `documentation/`, by putting an `index.html`
+-- within, or support `documentation`, by making a `documentation.html` file
+-- instead.
+webAppLookup :: ETagLookup -> FilePath -> Pieces -> IO LookupResult
+webAppLookup hashFunc prefix pieces = fileHelperLR hashFunc fp lastPiece
+ where
+  fp = pathFromPieces prefix pieces'
+  pieces' = initPieces ++ [lastPiece]
+  (initPieces, lastPiece) | null pieces = ([], unsafeToPiece "index.html")
+                          | Just (last pieces) == toPiece "" = (init pieces, unsafeToPiece "index.html")
+                          | otherwise   =
+                              let lastP = case fromPiece (last pieces) of
+                                    s | T.isSuffixOf ".txt" s -> last pieces
+                                    s -> unsafeToPiece $ s <> ".html"
+                              in (init pieces, lastP)
```

Comment:
Maybe the use of `unsafe` functions can be avoided here? And we can use a sum-type of `LookupResult` to more clearly represent we what we'd like. 

eg: 

``` haskell 

data LookupResult = UseIndexHtml | FoundFile Piece | NotFound
  deriving Show 
```

We can then better represent what the result of a lookup was.